### PR TITLE
Add a multilabel field

### PR DESCRIPTION
**Pull request by ibeltagy**
Adding a multilabel field which is useful for multi-label multi-class classification. 
Notes:
- I don't like how I am reading the vocabulary size, but I didn't find a better way. I am open to better suggestions. 
- The `as_tensor` function returns a dense vector of zeros and ones. It would have been more efficient to work with a list of indices instead, but that's not supported by torch.nn.BCEWithLogitsLoss

### CODE REVIEW

File: allennlp/data/fields/multilabel_field.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    # This warning will be repeated for every instantiation of this class (i.e for every data
+    # instance), spewing a lot of warnings so this class variable is used to only log a single
+    # warning per namespace.
+    _already_warned_namespaces: Set[str] = set()
+
+    # Dictionary from a vocab namespace to size. Need the vocab size to map labels to a tensor
+    _vocab_size: Dict[str, int] = defaultdict(int)
+
+    def __init__(self,
+                 labels: Set[Union[str, int]],
```
Comment (joelgrus) [inline allennlp/data/fields/multilabel_field.py]: I think this should be `Sequence` not `Set`, so that it type-checks to pass in a list (as you do in some places)

File: allennlp/data/fields/multilabel_field.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        self._maybe_warn_for_namespace(label_namespace)
+
+        if skip_indexing:
+            for label in labels:
+                if not isinstance(label, int):
+                    raise ConfigurationError("In order to skip indexing, your labels must be integers. "
+                                             "Found label = {}".format(label))
+                # vocabulary size = largest label id
+                MultiLabelField._vocab_size[self._label_namespace] = \
+                    max(label + 1, MultiLabelField._vocab_size[self._label_namespace])
```
Comment (joelgrus) [inline allennlp/data/fields/multilabel_field.py]: you don't have to do this change, but IMO it would be a little more readable if this wasn't a for loop

```
if skip_indexing:
    if not all(isinstance(label, int) for label in labels):
        raise ...
    largest_label_id = max(labels)
    MultiLabelField._vocab_size[..] = max(largest_label_id + 1, ...)
```

File: allennlp/data/fields/multilabel_field.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+            self._label_ids = labels
+        else:
+            for label in labels:
+                if not isinstance(label, str):
```
Comment (joelgrus) [inline allennlp/data/fields/multilabel_field.py]: similarly, I would probably just use `all()` here

File: allennlp/data/fields/multilabel_field.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                    raise ConfigurationError("MultiLabelFields expects string labels if skip_indexing=False. "
+                                             "Found label: {} with type: {}.".format(label, type(label)))
+
+    def _maybe_warn_for_namespace(self, label_namespace: str) -> None:
```
Comment (joelgrus) [inline allennlp/data/fields/multilabel_field.py]: it looks like this is only ever called in the constructor, so I might just put the logic there rather than have a separate function for it? (and then you can just use `label_namespace` instead of `self._label_namespace` which would make it a tiny bit easier to read

File: allennlp/data/fields/multilabel_field.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+class MultiLabelField(Field[torch.Tensor]):
+    """
+    A ``MultiLabelField`` is an extension of the :class:`LabelField` that allows for multiple labels.
+    It is particularly useful in multi-label classification where more than one label can be correct.
+    As with the :class:`LabelField`, labels are either strings of text or 0-indexed integers (if you wish
+    to skip indexing by passing skip_indexing=True).
+    If the labels need indexing, we will use a :class:`Vocabulary` to convert the string labels
+    into integers.
+
+    This field will get converted into a vector of length equals the vocabulary size with
```
Comment (joelgrus) [inline allennlp/data/fields/multilabel_field.py]: of length equal to

File: allennlp/data/fields/multilabel_field.py (hunk 1 above)
Comment (ibeltagy) [reply allennlp/data/fields/multilabel_field.py]: done

File: allennlp/data/fields/multilabel_field.py (hunk 2 above)
Comment (ibeltagy) [reply allennlp/data/fields/multilabel_field.py]: good point. Done

File: allennlp/data/fields/multilabel_field.py (hunk 3 above)
Comment (ibeltagy) [reply allennlp/data/fields/multilabel_field.py]: done.

File: allennlp/data/fields/multilabel_field.py (hunk 5 above)
Comment (ibeltagy) [reply allennlp/data/fields/multilabel_field.py]: fixed

File: allennlp/data/fields/multilabel_field.py (hunk 4 above)
Comment (ibeltagy) [reply allennlp/data/fields/multilabel_field.py]: good point, but how about keeping it as a function just to remain consistent with the code in allennlp/data/fields/label_field.py ?

**Comment by matt-gardner** [comment]
We use to do this in all of our fields.  There's a pretty easy way to do it, like this: https://github.com/allenai/allennlp/blob/048cba07cf94d43376092b61b7e50500bb7f928d/allennlp/data/fields/label_field.py#L61-L75

**Comment by DeNeutoy** [comment]
Yep, we settled on that.

**Comment by ibeltagy** [comment]
Replaced the static variable with something like what @matt-gardner and @DeNeutoy suggested above. Problem with this is that nothing checks if the vocabulary size changed or not, but things will break anyways if the vocabulary size changes.

File: allennlp/data/fields/multilabel_field.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        self._label_namespace = label_namespace
+        self._label_ids = None
+        self._maybe_warn_for_namespace(label_namespace)
+        self._num_labels = num_labels
+
+        if skip_indexing:
+            if not all(isinstance(label, int) for label in labels):
+                raise ConfigurationError("In order to skip indexing, your labels must be integers. "
+                                         "Found labels = {}".format(labels))
+            if not num_labels:
```
Comment (DeNeutoy) [inline allennlp/data/fields/multilabel_field.py]: This isn't necessary - just add an `if self._num_labels is not None:` block in `index`, because then you could also pass the number of labels if you know it.

File: allennlp/data/fields/multilabel_field.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        word).  If you have multiple different label fields in your data, you should make sure you
+        use different namespaces for each one, always using the suffix "labels" (e.g.,
+        "passage_labels" and "question_labels").
+    skip_indexing : ``bool``, optional (default=False)
+        If your labels are 0-indexed integers, you can pass in this flag, and we'll skip the indexing
+        step.  If this is ``False`` and your labels are not strings, this throws a ``ConfigurationError``.
+    num_labels : ``int``, optional (default=None)
+        If ``skip_indexing=True``, the total number of possible labels should be provided, which is needed
+        in deciding the size of the output tensor. `num_labels` should equal largest label id + 1
+        If ``skip_indexing=False``, it should remain None
```
Comment (DeNeutoy) [inline allennlp/data/fields/multilabel_field.py]: Full stop.

File: allennlp/data/fields/multilabel_field.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            if not all(cast(int, label) < num_labels for label in labels):
+                raise ConfigurationError("All labels should be < num_labels. "
+                                         "Found num_labels = {} and labels = {} ".format(num_labels, labels))
+
+            self._label_ids = labels
+        else:
+            if not all(isinstance(label, str) for label in labels):
+                raise ConfigurationError("MultiLabelFields expects string labels if skip_indexing=False. "
+                                         "Found labels: {}".format(labels))
+            if num_labels:
```
Comment (DeNeutoy) [inline allennlp/data/fields/multilabel_field.py]: This isn't necessary - just add an `if self._num_labels is not None:` block in `index`, because then you could also pass the number of labels if you know it.

**Comment by RamanRajarathinam** [comment]
As I see, I need to set 'hidden_dims' size in config.json before I start training MultiLableClassificaiton problem. 
Is there any solution to skip setting hidden_dims and finding from MultiLabelField object?

**Comment by matt-gardner** [comment]
@RamanRajarathinam, hidden dimension has to do with model size, not with data processing.  When constructing a model, there are typically some dimensions that you can compute from your inputs and some that you just need to be given.  You can look at our models to see examples of these different cases.  And if you have further questions, please open a separate issue to ask them.

**Comment by RamanRajarathinam** [comment]
@matt-gardner Sry If I wasn't clear last time. What did I mean by hidden_dims' size is, Its last layer size should be multi-label count. Since multi-label count cannot be determined at start of the process, is there any approach to determine the count value inside the program?

**Comment by matt-gardner** [comment]
The multi-label field uses the vocabulary to determine the number of labels: https://github.com/allenai/allennlp/blob/3cdb7e21c42aa5b3938cbd405cdc0f34f3178e9f/allennlp/data/fields/multilabel_field.py#L101-L102

In your model, you get passed the vocabulary, and you can get the number of labels in the same way, like we do for other classification tasks: https://github.com/allenai/allennlp/blob/3cdb7e21c42aa5b3938cbd405cdc0f34f3178e9f/allennlp/models/crf_tagger.py#L83
